dask_bigquery/core.py

```diff
@@ -0,0 +1,231 @@
+        fields = (partition_field, *fields)
+
+    # These read tasks seems to cause deadlocks (or at least long stuck workers out of touch with
+    # the scheduler), particularly when mixed with other tasks that execute C code. Anecdotally
+    # annotating the tasks with a higher priority seems to help (but not fully solve) the issue at
+    # the expense of higher cluster memory usage.
+    with bigquery_client(project_id, with_storage_api=True) as (
+        bq_client,
+        bqs_client,
+    ), dask.annotate(priority=1):
+        table_ref = bq_client.get_table(".".join((dataset_id, table_id)))
+        if table_ref.table_type == "VIEW":
+            raise TypeError("Table type VIEW not supported")
+
+        # The protobuf types can't be pickled (may be able to tweak w/ copyreg), so instead use a
+        # generator func.
+        def make_create_read_session_request(row_filter=""):
+            return bigquery_storage.types.CreateReadSessionRequest(
+                max_stream_count=100,  # 0 -> use as many streams as BQ Storage will provide
+                parent=f"projects/{project_id}",
+                read_session=bigquery_storage.types.ReadSession(
+                    data_format=bigquery_storage.types.DataFormat.ARROW,
+                    read_options=bigquery_storage.types.ReadSession.TableReadOptions(
+                        row_restriction=row_filter,
+                        selected_fields=fields,
+                    ),
+                    table=table_ref.to_bqstorage(),
+                ),
+            )
+
+        # Create a read session in order to detect the schema.
+        # Read sessions are light weight and will be auto-deleted after 24 hours.
+        session = bqs_client.create_read_session(
+            make_create_read_session_request(row_filter=row_filter)
+        )
+        schema = pyarrow.ipc.read_schema(
+            pyarrow.py_buffer(session.arrow_schema.serialized_schema)
+        )
+        meta = schema.empty_table().to_pandas()
+        delayed_kwargs = dict(prefix=f"{dataset_id}.{table_id}-")
+
+        if partition_field is not None:
+            if row_filter:
+                raise ValueError("Cannot pass both `partition_field` and `row_filter`")
+            delayed_kwargs["meta"] = meta.set_index(partition_field, drop=True)
+
+            if partitions is None:
+                logging.info(
+                    "Specified `partition_field` without `partitions`; reading full table."
+                )
+                partitions = pd.read_gbq(
+                    f"SELECT DISTINCT {partition_field} FROM {dataset_id}.{table_id}",
+                    project_id=project_id,
+                )[partition_field].tolist()
+                # TODO generalize to ranges (as opposed to discrete values)
+
+            partitions = sorted(partitions)
+            delayed_kwargs["divisions"] = (*partitions, partitions[-1])
```

Comment:
@bnaul I noticed in the example I run, that this line causes to have the last partition to contain only 1 element, but that element could have fit into the previous to last partition. What is the reason you separate the last partition?

Reply:
not sure why it's not working correctly for you but the idea is that you need n+1 divisions for n partitions. seems to work OK here
```
import dask.dataframe as dd
from dask import delayed

@delayed
def make_df(d):
    return pd.DataFrame({"date": d, "x": np.random.random(10)}).set_index("date")

dates = pd.date_range("2020-01-01", "2020-01-08")
ddf = dd.from_delayed([make_df(d) for d in dates], divisions=[*dates, dates[-1]])
ddf.map_partitions(len).compute()

Out[62]:
0    10
1    10
2    10
3    10
4    10
5    10
6    10
7    10
dtype: int64
```

Reply:
I wonder if it's related to how the data is originally partitioned. For example when I read one of the tables of the covid public data set that I copied on "my_project" I see this

```python
from dask_bigquery import read_gbq

ddf= read_gbq(
                project_id="my_project",
                dataset_id="covid19_public_forecasts",
                table_id="county_14d",)

ddf.map_partitions(len).compute()
```
Notice the last two partitions...
```
0     3164
1     3164
2     3164
3     3164
4     3164
5     3164
6     3164
7     3164
8     3164
9     3164
10    3164
11    3164
12    3164
13    3164
14    3164
15    3164
16    3164
17    3164
18    3164
19    3164
20    3164
21    3164
22    3164
23    3164
24    3164
25    3164
26    3164
27    3164
28    3164
29    3164
30    3164
31    3164
32    3164
33    3164
34    3164
35    3164
36    3164
37    3164
38    3164
39    3164
40    3164
41    3163
42       1
dtype: int64
```